Capture request timeout error and increase the timeout value

oeway opened the pull request:
This PR aims to fix https://github.com/jupyterhub/jupyter-server-proxy/issues/182, it:
1. Increase the `connect_timeout=250.0, request_timeout=300.0` (both defaults to 20s) (these two values are used as default for FireFox).
2. Capture the timeout error, return `599: Timeout during request` rather than `500: Internal Server Error`.

ryanlovett commented:
Thanks for investigating and the PR! Your logic seems sound to me. Just one comment...

I'd prefer if we catch tornado.httpclient.HTTPClientError, check whether its .code is a 599, and respond in the way you've written, raising if otherwise. Tornado chooses the simple client by default, but we us the base httpclient interface all throughout and it is common to both tornado's simple and curl http clients. (not that we provide a way to use the curl client instead)

What do you think? Anyone else have a comment?

oeway commented:
@ryanlovett Although I am not familiar with the project enough, I think that make sense and I have changed according to your suggestion.

In the meantime, I did a bit more investigation on the connection timeout for browsers, Firefox (on Mac) for example has the following settings (can be changed via `about:config`):
`network.http.response.timeout=300`
`network.http.connection-timeout=90`
`network.http.connection-retry-timeout=250`

Thus, I changed the `connect_timeout` for the HttpClient to 250s, and also increased the `request_timeout` to 300s. I think it's safe and better to use a large value for both, since we don't want the proxy client to give up a lot earlier than the browser.

betatim commented on jupyter_server_proxy/handlers.py:
```diff
@@ -213,7 +213,18 @@ def _check_host_whitelist(self, host):
+
+        try:
+            response = await client.fetch(req, raise_error=False)
+        except httpclient.HTTPError as err:
```
The `client.fetch` call uses `raise_error=False`. I think this means no exceptions will be raised. Is this correct or is the argument confusingly named (no exceptions except for timeouts)?

If there are no exceptions we should probably deal with timeouts in the code below

oeway replied:
We do see the timeout exception even with `raise_error=False`, see the stack trace here: https://github.com/jupyterhub/jupyter-server-proxy/issues/182.

In the [docs](https://www.tornadoweb.org/en/stable/httpclient.html#tornado.httpclient.AsyncHTTPClient.fetch), it said explicitly it will not suppress all errors:
```
The raise_error=False argument only affects the HTTPError raised when a non-200 response code is used, instead of suppressing all errors.
```

So why the timeout error is not `HTTPError` and being suppressed? I guess if there is a timeout, then there won't be any response, thus the error cannot be suppressed, I think.

betatim replied:
Aha! Thanks for digging out the docs :)

Maybe we can add a comment to the code which contains your response/link to docs to remind future generations that this seemingly weird setup is on purpose.

oeway replied:
Good idea, done!

ryanlovett replied:
@betatim I think raise_error=False tells the client to return the response to the caller, regardless of error code, rather than raise an exception. I think we want to always get the response, but we could conceivably return other things to the user based on the http code if we wanted.

Also, in tornado <= 5.1 it would [set the return code for errors to 599](https://www.tornadoweb.org/en/branch5.1/httpclient.html#tornado.httpclient.AsyncHTTPClient.fetch). If people are still using 5.1 with this proxy and this PR, they may start to see a lot of timeout errors whereas before we would just return the response without checking the error code. I don't think we care too much but it's worth noting. If a lot of people are still on 5.1, I think we'd have to insert a check on the tornado version and use the old behavior for <= 5.1.

oeway replied:
@ryanlovett Could you please clarify why `they may start to see a lot of timeout errors whereas before we would just return the response without checking the error code.`?  

To be clear, the only thing we did in this PR is to capture the error which previously treated as internal server error (we are not changing the setting of `raise_error=False`). In the link you mentioned, it said:
> The raise_error=False argument currently suppresses all errors, encapsulating them in HTTPResponse objects with a 599 response code. This will change in Tornado 6.0: raise_error=False will only affect the HTTPError raised when a non-200 response code is used.

If I understand correctly, tornado <=5.1 won't raise any error, so the code should function exactly as before. And for tornado >5.1 (I am testing with 6.0+), it will raise the error then we will start to return timeout errors instead of saying the internal server error.

ryanlovett replied:
We won't see more exceptions due to raise_error=False, but when suppressing the exceptions, tornado <= 5.1 would set the return code to be 599. This PR examines that code... only when there is an exception. So... never mind. Sorry.

rsignell-usgs replied:
This is exciting guys!  Can't wait to try this out!  Thanks @oeway!

ryanlovett replied:
@betatim Any final thoughts before merging?

betatim replied:
No thoughts. My confusion was from the "split" behaviour of the new meaning of the argument. That was addressed in https://github.com/jupyterhub/jupyter-server-proxy/pull/183#discussion_r398504580. Marking this conversation as resolved.

betatim commented:
Thanks for working on this and the education about the new behaviour of the "no exceptions please" flag :)
